src/datasets/formatting/np_formatter.py

```diff
@@ -0,0 +1,80 @@
+import sys
+
+import numpy as np
+import pyarrow as pa
+
+from .. import config
+from ..utils.py_utils import map_nested
+from .formatting import Formatter
+
+
+class NumpyFormatter(Formatter[dict, np.ndarray, dict]):
+    def __init__(self, features=None, decoded=True, **np_array_kwargs):
+        super().__init__(features=features, decoded=decoded)
+        self.np_array_kwargs = np_array_kwargs
+
+    def _consolidate(self, column):
+        if isinstance(column, list):
+            if column and all(
+                isinstance(x, np.ndarray) and x.shape == column[0].shape and x.dtype == column[0].dtype for x in column
+            ):
+                return np.stack(column)
+            else:
+                return np.array(column, dtype=object)
```

Comment:
Casting to `np.object` can sometimes fail (see https://stackoverflow.com/q/51005699), so I think we need to wrap this with try-except and return a list if that happens

Reply:
I used
```python
out = np.empty(len(column), dtype=object)
out[:] = column
```
which is suggested in the SO answer, and it works well :)

let me know if it's all good for you